### PR TITLE
Fixed compile error for --enable-debug on non-Linux

### PR DESCRIPTION
**Pull request by cculianu**
On BSD and Darwin and other non-Linux, `--enable-debug` compilation
was producing the following error:

```

  sync.cpp:36:27: error: no member named 'lexical_cast' in namespace 'boost'
      uint64_t tid = boost::lexical_cast<uint64_t>(boost::this_thread::get_id());
                   ~~~~~~~^
  sync.cpp:36:40: error: unexpected type name 'uint64_t': expected expression
      uint64_t tid = boost::lexical_cast<uint64_t>(boost::this_thread::get_id());
                                       ^
  2 errors generated.

```

Additionally, I am not even sure the `#include <sys/syscall.h>` would be kosher with 
Windows compilation either in this case.

I tried fixing this by `#include`ing the requisite lexical_cast header for
boost, but that produced a runtime error due to an uncaught exception on
Darwin (bad_lexical_cast) -- presumably the thread-id on Darwin cannot
be lexically cast to int.

This is not surprising: There is no guaranteed way to turn a std::thread::id 
(or boost::thread::id) into an int save for hashing it using std::hash.  So we do that 
for the non-linux case, and compilation succeeds and bitcoind runs ok.

### CODE REVIEW

**Comment by ptschip** [comment]
Travis not kicking off on this PR either.  Need to re-enable the account via travis support.

**Review by sickpig — APPROVED** [review]

**Comment by cculianu** [comment]
> Travis not kicking off on this PR either. Need to re-enable the account via travis support.

Contacting Travis now.. Thanks.

**Comment by cculianu** [comment]
Ok, I contacted Travis. Hopefully they will get back to me soon. Sorry about that! I had no idea I was flagged!

**Comment by ptschip** [comment]
no worries, they blanket flagged everybody for some reason.

**Comment by cculianu** [comment]
Note this PR #1857 touches this code too and moves it elsewhere; but as yet it has not completely fixed the problems with it for Darwin... I wrote a note there to tell the PR author about Darwin issues...

**Comment by cculianu** [comment]
> no worries, they blanket flagged everybody for some reason.

FYI, for the curious -- apparently they flag anything to do with `bitcoind` or bitcoin in general:

### From Travis Support

> Hey Calin,
> 
> Thanks for the email and apologies for the issue.
> 
> This is a part of our abuse prevention system, which blocks a repo if it looks like someone is trying to mine cryptocurrency. It is an evolving system, as we work to ensure our service is reliable and performant for the community and our customers. We make every attempt to avoid false positives, but they do happen, and we know this can cause disruptions in your work. Could I ask how you're using Travis for this repository, and how you're using bitcoind in relation to Travis?
> 
> Look forward to hearing from you.
> 
> Best,
> Sabri
>

**Comment by Greg-Griffith** [comment]
ill merge this fix into my PR via rebase once this PR is merged. thanks for finding it. 

This function's entire purpose is to be used with gbd to make some debugging easier. so if the hashing cast produces weird results i wouldnt worry too much about it.

**Pull request merged**
